src/state-item.c

```diff
@@ -0,0 +1,417 @@
+                    state_item_number dstSI = state_item_index_lookup(dst->number, k);
+                    
+                    trans[j] = dstSI;
+                    bitset_set(rev_trans[dstSI], j);
+                    break;
+                }
+            }
+        }
+    }
+}
+
+bitset
+prods_lookup(state_item_number si)
+{
+    return hash_pair_lookup(prods, si);
+}
+
+bitset
+rev_prods_lookup(state_item_number si)
+{
+    return hash_pair_lookup(rev_prods, si);
+}
+
+/**
+ * Initialize prods and revProds maps.
+ */
+static void init_prods()
+{
+    prods = hash_initialize (nstate_items,
+                             NULL,
+                             hash_pair_hasher,
+                             hash_pair_comparator,
+                             NULL);
+    rev_prods = hash_initialize (nstate_items,
+                                 NULL,
+                                 hash_pair_hasher,
+                                 hash_pair_comparator,
+                                 NULL);
+    for (int i = 0; i < nstates; i++) {
+        state *state = states[i];
+        // closure_map is a hash map from nonterminals to a set
+        // of the items that produce those nonterminals
+        Hash_table *closure_map = hash_initialize (nsyms - ntokens,                                                NULL,
+                                                 hash_pair_hasher,
+                                                 hash_pair_comparator,
+                                                 NULL);
+        
+        // Add the nitems of state to skip to the production portion
+        // of that state's state_items
+        for (int j = state_item_map[i] + state->nitems; j < state_item_map[i + 1]; j++) {
+            state_item *src = state_items + j;
+            item_number *item = src->item;
+            symbol_number lhs = item_rule(item)->lhs->number;
+            bitset itms = hash_pair_lookup(closure_map, lhs);
+            if (itms == NULL) {
+                itms = bitset_create(nstate_items, BITSET_SPARSE);
+                hash_pair_insert(closure_map, lhs, itms);
+            }
+            bitset_set(itms, j);
+        }
+        // For each item with a dot followed by a nonterminal,
+        // try to create a production edge.
+        for (int j = state_item_map[i]; j < state_item_map[i + 1]; j++) {
+            state_item *src = state_items + j;
+            item_number item = *(src->item);
+            // Skip reduce items and items with terminals after the dot
+            if (item_number_is_rule_number(item) || ISTOKEN(item)) continue;
+            symbol_number sym = item_number_as_symbol_number(item);
+            bitset lb = hash_pair_lookup(closure_map, sym);
+            if (lb != NULL) {
+                hash_pair *prod_hp = xmalloc(sizeof(hash_pair));
+                prod_hp->key = j;
+                prod_hp->l = lb;
+                //update prods
+                if (!hash_insert(prods, prod_hp))
```

Comment:
We should introduce and use `hash_xinsert` instead of having to call `xalloc_die` by hand everywhere.